src/coord/src/coord.rs

```diff
@@ -2958,65 +2981,67 @@ impl Coordinator {
-                    // Add the used sources and indexes to the recorded ids.
-                    timedomain_ids.extend(&determined.storage_ids);
-                    timedomain_ids.extend(&determined.compute_ids);
+                    let timestamp = self.determine_timestamp(
+                        session,
+                        &storage_ids,
+                        &compute_ids,
+                        PeekWhen::Immediately,
+                    )?;
                     let read_holds = transaction_holds::ReadHolds {
-                        time: determined.timestamp,
-                        storage_ids: determined.storage_ids,
-                        compute_ids: determined.compute_ids,
+                        time: timestamp,
+                        storage_ids,
+                        compute_ids,
                         compute_instance: DEFAULT_COMPUTE_INSTANCE_ID,
                     };
                     self.acquire_read_holds(&read_holds).await;
                     let txn_reads = TxnReads {
                         timestamp_independent,
-                        timedomain_ids: timedomain_ids.into_iter().collect(),
                         read_holds,
                     };
                     self.txn_reads.insert(conn_id, txn_reads);
-
-                    determined.timestamp
+                    timestamp
                 }
             };
             session.add_transaction_ops(TransactionOps::Peeks(timestamp))?;
 
-            // Verify that the references and indexes for this query are in the current
-            // read transaction.
-            let mut stmt_ids = HashSet::new();
-            stmt_ids.extend(source_ids.iter().collect::<HashSet<_>>());
-            // Using nearest_indexes here is a hack until #8318 is fixed. It's used because
-            // that's what determine_timestamp uses.
-            stmt_ids.extend(
-                self.catalog
-                    .nearest_indexes(&source_ids)
-                    .0
-                    .into_iter()
-                    .collect::<HashSet<_>>(),
-            );
+            // Verify that the references and indexes for this query are in the
+            // current read transaction.
+            //
+            // Using nearest_indexes here is a hack until #8318 is fixed. It's
+            // used because that's what determine_timestamp uses.
+            let (compute_ids, storage_ids) = self.catalog.nearest_indexes(&source_ids);
+            let storage_ids = BTreeSet::from_iter(storage_ids);
+            let compute_ids = BTreeSet::from_iter(compute_ids);
+
             let read_txn = self.txn_reads.get(&conn_id).unwrap();
+            let allowed_storage_ids =
+                BTreeSet::from_iter(read_txn.read_holds.storage_ids.iter().copied());
+            let allowed_compute_ids =
+                BTreeSet::from_iter(read_txn.read_holds.compute_ids.iter().copied());
+
```

Comment:
Lots of one-off `BTreeSet` allocation here. Should we do something smarter? Hold `BTreeSet`s in `TxnReadHold`? Perform set difference on `Vec`s directly?